File: src/seed/seed-interactions.ts
/**
 * seed-interactions.ts
 *
 * Generates realistic simulated interaction records for PR agents.
 * This script is ADDITIVE — it does NOT clear any existing collections.
 *
 * Usage:
 *   npm run seed:interactions          # seed ~300 records
 *   npm run seed:interactions --dry-run # preview counts only
 *
 * After seeding, run:
 *   npm run backfill:interactions      # fill interactionEmbedding via Voyage AI
 */

import dotenv from "dotenv";
dotenv.config();

import { getDb, closeClient } from "../db.js";
import { collections } from "../modeling/indexes.js";
import { buildInteractionEmbeddingText } from "../web/embedding.js";
import type {
  PatronInteractionRecord,
  InteractionType,
  PatronTier,
} from "../types.js";

// ---------- Config ----------

const isDryRun = process.argv.includes("--dry-run");

// Min/max interactions per PR agent
const MIN_PER_PR = 8;
const MAX_PER_PR = 20;

// How far back (days) to scatter interaction dates
const HISTORY_DAYS = 180;

// Cosine similarity threshold we want to pass for KPI searches — we engineer
// the detail fields so that the embedding text is semantically rich and
// distinct per group, giving realistic vector search results after backfill.

// ---------- Types ----------

type PatronStub = { patronId: string; tier: PatronTier; adt: number };
type PRStub    = { prAgentId: string; name: string; active: boolean; preferredTiers: PatronTier[] };

// ---------- Interaction type distribution per PR group ----------
// Groups are based on position in the sorted prAgentId list.
// This creates realistic variation so the KPI vector search shows
// clear differences between agents.

type TypeWeight = { type: InteractionType; weight: number };

function getTypeWeights(groupIndex: number, totalGroups: number): TypeWeight[] {
  const quarter = Math.floor((groupIndex / totalGroups) * 4);
  switch (quarter) {
    case 0:
      // Group A — Room & Transfer specialists (high-roller hosting)
      return [
        { type: "ROOM_COMP",    weight: 40 },
        { type: "TRANSFER",     weight: 30 },
        { type: "FB_COMP",      weight: 10 },
        { type: "OUTREACH",     weight: 10 },
        { type: "EVENT_INVITE", weight: 5  },
        { type: "REBATE",       weight: 5  },
      ];
    case 1:
      // Group B — Outreach & Event specialists (patron retention)
      return [
        { type: "OUTREACH",     weight: 40 },
        { type: "EVENT_INVITE", weight: 30 },
        { type: "FB_COMP",      weight: 10 },
        { type: "TRANSFER",     weight: 10 },
        { type: "ROOM_COMP",    weight: 5  },
        { type: "REBATE",       weight: 5  },
      ];
    case 2:
      // Group C — Rebate & F&B specialists (volume reward)
      return [
        { type: "REBATE",       weight: 35 },
        { type: "FB_COMP",      weight: 30 },
        { type: "OUTREACH",     weight: 15 },
        { type: "EVENT_INVITE", weight: 10 },
        { type: "ROOM_COMP",    weight: 5  },
        { type: "TRANSFER",     weight: 5  },
      ];
    default:
      // Group D — Generalist (balanced across all types)
      return [
        { type: "ROOM_COMP",    weight: 17 },
        { type: "FB_COMP",      weight: 17 },
        { type: "REBATE",       weight: 17 },
        { type: "EVENT_INVITE", weight: 17 },
        { type: "OUTREACH",     weight: 16 },
        { type: "TRANSFER",     weight: 16 },
      ];
  }
}

function pickWeightedType(weights: TypeWeight[]): InteractionType {
  const total = weights.reduce((s, w) => s + w.weight, 0);
  let r = Math.random() * total;
  for (const w of weights) {
    r -= w.weight;
    if (r <= 0) return w.type;
  }
  return weights[weights.length - 1].type;
}

// ---------- Value ranges by type ----------

function randomInt(min: number, max: number) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

function randomFrom<T>(arr: readonly T[]): T {
  return arr[Math.floor(Math.random() * arr.length)];
}

function randomDate(daysBack: number): Date {
  const ms = Math.random() * daysBack * 24 * 60 * 60 * 1000;
  return new Date(Date.now() - ms);
}

function buildDetail(
  type: InteractionType,
  tier: PatronTier
): { detail: PatronInteractionRecord["detail"]; totalValueHKD: number } {
  switch (type) {
    case "ROOM_COMP": {
      const isPlatDiamond = tier === "Platinum" || tier === "Diamond";
      const roomTypes = isPlatDiamond
        ? ["Grand Suite", "Presidential Suite", "VIP Harbour Suite"]
        : ["Superior Room", "Deluxe Room", "Premier Room"];
      const roomType  = randomFrom(roomTypes);
      const roomNights = randomInt(1, isPlatDiamond ? 3 : 2);
      const nightRate  = isPlatDiamond ? randomInt(6000, 15000) : randomInt(2000, 6000);
      const roomValue  = roomNights * nightRate;
      return {
        detail: { roomType, roomNights, roomValue },
        totalValueHKD: roomValue,
      };
    }
    case "FB_COMP": {
      const venues = ["The Starlight Restaurant", "Dynasty Cantonese", "La Maison Brasserie", "Jade Garden", "Skyview Lounge"];
      const venue    = randomFrom(venues);
      const fbAmount = randomInt(800, 6000);
      return { detail: { venue, fbAmount }, totalValueHKD: fbAmount };
    }
    case "REBATE": {
      const rebateRate   = randomFrom([0.005, 0.008, 0.010, 0.012, 0.015, 0.018, 0.020]);
      const rebateAmount = randomInt(3000, 25000);
      return { detail: { rebateRate, rebateAmount }, totalValueHKD: rebateAmount };
    }
    case "EVENT_INVITE": {
      const events = [
        "8月 Diamond VIP 晚宴",
        "中秋貴賓品酒會",
        "新年 Platinum 專屬招待",
        "週年慶 VIP 音樂盛典",
        "高爾夫球友誼賽",
        "私人藝術品鑑會",
      ];
      const eventName = randomFrom(events);
      const attended  = Math.random() > 0.25;
      return { detail: { eventName, attended }, totalValueHKD: randomInt(500, 3000) };
    }
    case "OUTREACH": {
      const channels: Array<"Phone" | "In-Person" | "WeChat" | "WhatsApp"> = [
        "Phone", "WeChat", "WhatsApp", "In-Person",
      ];
      const outcomes: Array<"Positive" | "Neutral" | "No Answer" | "Declined"> = [
        "Positive", "Positive", "Positive", "Neutral", "No Answer", "Declined",
      ];
      const channel = randomFrom(channels);
      const outcome = randomFrom(outcomes);
      const noteOptions = [
        "確認下月回訪計劃",
        "賭客表示近期會帶朋友前來",
        "已安排下次桌位預留",
        "詢問最新優惠詳情",
        "賭客對本次服務表示滿意",
        "跟進上次 VIP 活動反饋",
      ];
      const notes = outcome !== "No Answer" ? randomFrom(noteOptions) : undefined;
      return { detail: { channel, outcome, notes }, totalValueHKD: 0 };
    }
    case "TRANSFER": {
      const transferTypes: Array<"Airport" | "Hotel" | "Venue"> = ["Airport", "Hotel", "Venue"];
      const vehicleClasses: Array<"Standard" | "Luxury"> =
        tier === "Platinum" || tier === "Diamond" ? ["Luxury"] : ["Standard", "Luxury"];
      const transferType  = randomFrom(transferTypes);
      const vehicleClass  = randomFrom(vehicleClasses);
      const value = vehicleClass === "Luxury" ? randomInt(800, 2500) : randomInt(300, 800);
      return { detail: { transferType, vehicleClass }, totalValueHKD: value };
    }
  }
}

// ---------- Build interaction record ----------

function buildRecord(
  prAgentId: string,
  patron: PatronStub,
  type: InteractionType,
  now: Date
): PatronInteractionRecord {
  const occurredAt = randomDate(HISTORY_DAYS);
  const { detail, totalValueHKD } = buildDetail(type, patron.tier);

  // Build the embedding text now (stored as a plain field so backfill can
  // regenerate without recomputing) — embedding vector itself is NOT stored here
  const interactionId = `INT-${now.getTime()}-${prAgentId}-${patron.patronId}-${Math.random().toString(36).slice(2, 7)}`
    .replace(/[^A-Z0-9-]/gi, "-")
    .toUpperCase();

  return {
    interactionId,
    patronId:         patron.patronId,
    type,
    detail,
    totalValueHKD,
    occurredAt,
    recordedBy:       prAgentId,
    recordedAt:       now,
    patronTierAtTime: patron.tier,
    patronAdtAtTime:  patron.adt,
    // interactionEmbedding is intentionally omitted — run backfill:interactions to fill
  };
}

// ---------- Patron pool helpers ----------

function buildTierPatronMap(patrons: PatronStub[]): Map<PatronTier, PatronStub[]> {
  const map = new Map<PatronTier, PatronStub[]>();
  for (const p of patrons) {
    if (!map.has(p.tier)) map.set(p.tier, []);
    map.get(p.tier)!.push(p);
  }
  return map;
}

/**
 * Pick a patron for a PR, biased toward tiers that PR prefers.
 * Falls back to any patron if preferred tier pool is empty.
 */
function pickPatron(
  preferredTiers: PatronTier[],
  tierMap: Map<PatronTier, PatronStub[]>,
  allPatrons: PatronStub[]
): PatronStub {
  // 70% chance to pick from preferred tier
  if (Math.random() < 0.7 && preferredTiers.length > 0) {
    const tier = randomFrom(preferredTiers);
    const pool = tierMap.get(tier);
    if (pool && pool.length > 0) return randomFrom(pool);
  }
  return randomFrom(allPatrons);
}

// ---------- Main ----------

async function seedInteractions() {
  const db  = await getDb();
  const now = new Date();

  // 1. Load existing PR agents
  const prDocs = await db
    .collection(collections.prAgents)
    .find({}, { projection: { prAgentId: 1, name: 1, active: 1, preferredTiers: 1 } })
    .sort({ prAgentId: 1 })
    .toArray() as unknown as PRStub[];

  if (prDocs.length === 0) {
    console.error("No PR agents found. Run `npm run seed` first.");
    process.exitCode = 1;
    return;
  }

  // 2. Load existing patrons (take up to 300 for performance)
  const patronDocs = await db
    .collection(collections.patrons)
    .find({}, { projection: { patronId: 1, tier: 1, adt: 1 } })
    .limit(300)
    .toArray() as unknown as PatronStub[];

  if (patronDocs.length === 0) {
    console.error("No patrons found. Run `npm run seed` first.");
    process.exitCode = 1;
    return;
  }

  const tierMap = buildTierPatronMap(patronDocs);

  // 3. Generate records for each PR
  const allRecords: PatronInteractionRecord[] = [];

  prDocs.forEach((pr, idx) => {
    const count      = randomInt(MIN_PER_PR, MAX_PER_PR);
    const typeWeights = getTypeWeights(idx, prDocs.length);

    for (let i = 0; i < count; i++) {
      const type   = pickWeightedType(typeWeights);
      const patron = pickPatron(pr.preferredTiers ?? [], tierMap, patronDocs);
      allRecords.push(buildRecord(pr.prAgentId, patron, type, now));
    }
  });

  // 4. Preview or insert
  if (isDryRun) {
    // Summarise by PR
    const byPr: Record<string, Record<string, number>> = {};
    for (const r of allRecords) {
      if (!byPr[r.recordedBy]) byPr[r.recordedBy] = {};
      byPr[r.recordedBy][r.type] = (byPr[r.recordedBy][r.type] ?? 0) + 1;
    }
    console.log(JSON.stringify({ dryRun: true, totalRecords: allRecords.length, byPr }, null, 2));
    return;
  }

  await db
    .collection(collections.patronInteractions)
    .insertMany(allRecords as never[]);

  // 5. Summary
  const typeBreakdown: Record<string, number> = {};
  for (const r of allRecords) {
    typeBreakdown[r.type] = (typeBreakdown[r.type] ?? 0) + 1;
  }
  console.log(`\n✓ Inserted ${allRecords.length} interaction records across ${prDocs.length} PR agents.`);
  console.log(`  Type breakdown:`, typeBreakdown);
  console.log(`  Date range: past ${HISTORY_DAYS} days`);
  console.log(`\nNext step: run \`npm run backfill:interactions\` to generate embeddings for KPI vector search.\n`);
}

seedInteractions()
  .catch((err) => {
    console.error("seed-interactions failed:", err);
    process.exitCode = 1;
  })
  .finally(closeClient);

File: server/app/seed/seed_interactions.py
"""Seed ``patron_interaction_history`` with realistic PR-agent interactions.

Port of ``src/seed/seed-interactions.ts``. ADDITIVE — does NOT clear the
collection. After seeding, run
``uv run python -m app.scripts.backfill_interaction_embeddings`` to fill
``interactionEmbedding``.

Run with::

    uv run python -m app.seed.seed_interactions              # seed
    uv run python -m app.seed.seed_interactions --dry-run    # preview
"""

from __future__ import annotations

import asyncio
import json
import random
import sys
from datetime import datetime, timedelta, timezone
from typing import Any

from app.collections import web_collections as cols
from app.db import close_client, get_db

# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------

MIN_PER_PR = 8
MAX_PER_PR = 20
HISTORY_DAYS = 180


# ---------------------------------------------------------------------------
# Interaction type distribution per PR group
# ---------------------------------------------------------------------------


def _type_weights(group_index: int, total_groups: int) -> list[tuple[str, int]]:
    """Weighted interaction-type distribution based on the PR's group quarter."""
    quarter = int((group_index / total_groups) * 4) if total_groups > 0 else 0
    if quarter == 0:
        # Group A — Room & Transfer specialists (high-roller hosting)
        return [
            ("ROOM_COMP", 40),
            ("TRANSFER", 30),
            ("FB_COMP", 10),
            ("OUTREACH", 10),
            ("EVENT_INVITE", 5),
            ("REBATE", 5),
        ]
    if quarter == 1:
        # Group B — Outreach & Event specialists (patron retention)
        return [
            ("OUTREACH", 40),
            ("EVENT_INVITE", 30),
            ("FB_COMP", 10),
            ("TRANSFER", 10),
            ("ROOM_COMP", 5),
            ("REBATE", 5),
        ]
    if quarter == 2:
        # Group C — Rebate & F&B specialists (volume reward)
        return [
            ("REBATE", 35),
            ("FB_COMP", 30),
            ("OUTREACH", 15),
            ("EVENT_INVITE", 10),
            ("ROOM_COMP", 5),
            ("TRANSFER", 5),
        ]
    # Group D — Generalist
    return [
        ("ROOM_COMP", 17),
        ("FB_COMP", 17),
        ("REBATE", 17),
        ("EVENT_INVITE", 17),
        ("OUTREACH", 16),
        ("TRANSFER", 16),
    ]


def _pick_weighted_type(weights: list[tuple[str, int]]) -> str:
    total = sum(w for _, w in weights)
    r = random.random() * total
    for value, weight in weights:
        r -= weight
        if r <= 0:
            return value
    return weights[-1][0]


# ---------------------------------------------------------------------------
# Small helpers
# ---------------------------------------------------------------------------


def _rand_int(min_val: int, max_val: int) -> int:
    return random.randint(min_val, max_val)


def _rand_from(seq):
    return random.choice(seq)


def _random_date(days_back: int) -> datetime:
    now = datetime.now(timezone.utc)
    ms = random.random() * days_back * 24 * 60 * 60 * 1000
    return now - timedelta(milliseconds=ms)


# ---------------------------------------------------------------------------
# Detail builders
# ---------------------------------------------------------------------------


def _build_detail(
    interaction_type: str, tier: str | None
) -> tuple[dict[str, Any], float]:
    if interaction_type == "ROOM_COMP":
        is_plat_diamond = tier in ("Platinum", "Diamond")
        room_types = (
            ["Grand Suite", "Presidential Suite", "VIP Harbour Suite"]
            if is_plat_diamond
            else ["Superior Room", "Deluxe Room", "Premier Room"]
        )
        room_type = _rand_from(room_types)
        room_nights = _rand_int(1, 3 if is_plat_diamond else 2)
        night_rate = (
            _rand_int(6000, 15000) if is_plat_diamond else _rand_int(2000, 6000)
        )
        room_value = room_nights * night_rate
        return (
            {"roomType": room_type, "roomNights": room_nights, "roomValue": room_value},
            room_value,
        )

    if interaction_type == "FB_COMP":
        venues = [
            "The Starlight Restaurant",
            "Dynasty Cantonese",
            "La Maison Brasserie",
            "Jade Garden",
            "Skyview Lounge",
        ]
        venue = _rand_from(venues)
        fb_amount = _rand_int(800, 6000)
        return {"venue": venue, "fbAmount": fb_amount}, fb_amount

    if interaction_type == "REBATE":
        rebate_rate = _rand_from([0.005, 0.008, 0.010, 0.012, 0.015, 0.018, 0.020])
        rebate_amount = _rand_int(3000, 25000)
        return (
            {"rebateRate": rebate_rate, "rebateAmount": rebate_amount},
            rebate_amount,
        )

    if interaction_type == "EVENT_INVITE":
        events = [
            "8月 Diamond VIP 晚宴",
            "中秋貴賓品酒會",
            "新年 Platinum 專屬招待",
            "週年慶 VIP 音樂盛典",
            "高爾夫球友誼賽",
            "私人藝術品鑑會",
        ]
        event_name = _rand_from(events)
        attended = random.random() > 0.25
        return (
            {"eventName": event_name, "attended": attended},
            _rand_int(500, 3000),
        )

    if interaction_type == "OUTREACH":
        channels = ["Phone", "WeChat", "WhatsApp", "In-Person"]
        outcomes = ["Positive", "Positive", "Positive", "Neutral", "No Answer", "Declined"]
        channel = _rand_from(channels)
        outcome = _rand_from(outcomes)
        note_options = [
            "確認下月回訪計劃",
            "賭客表示近期會帶朋友前來",
            "已安排下次桌位預留",
            "詢問最新優惠詳情",
            "賭客對本次服務表示滿意",
            "跟進上次 VIP 活動反饋",
        ]
        notes = _rand_from(note_options) if outcome != "No Answer" else None
        detail: dict[str, Any] = {"channel": channel, "outcome": outcome}
        if notes is not None:
            detail["notes"] = notes
        return detail, 0

    if interaction_type == "TRANSFER":
        transfer_types = ["Airport", "Hotel", "Venue"]
        vehicle_classes = (
            ["Luxury"] if tier in ("Platinum", "Diamond") else ["Standard", "Luxury"]
        )
        transfer_type = _rand_from(transfer_types)
        vehicle_class = _rand_from(vehicle_classes)
        value = (
            _rand_int(800, 2500) if vehicle_class == "Luxury" else _rand_int(300, 800)
        )
        return (
            {"transferType": transfer_type, "vehicleClass": vehicle_class},
            value,
        )

    # Unknown type — mirrors TS behaviour (no explicit default, so return empty)
    return {}, 0


def _build_record(
    pr_agent_id: str, patron: dict[str, Any], interaction_type: str, now: datetime
) -> dict[str, Any]:
    occurred_at = _random_date(HISTORY_DAYS)
    detail, total_value_hkd = _build_detail(interaction_type, patron.get("tier"))

    raw_suffix = "".join(random.choices("abcdefghijklmnopqrstuvwxyz0123456789", k=5))
    interaction_id = (
        f"INT-{int(now.timestamp() * 1000)}-{pr_agent_id}-{patron['patronId']}-{raw_suffix}"
    ).upper()
    # TS strips anything outside [A-Z0-9-] (case-insensitive). All the pieces
    # above already satisfy that, so no substitution needed.

    return {
        "interactionId": interaction_id,
        "patronId": patron["patronId"],
        "type": interaction_type,
        "detail": detail,
        "totalValueHKD": total_value_hkd,
        "occurredAt": occurred_at,
        "recordedBy": pr_agent_id,
        "recordedAt": now,
        "patronTierAtTime": patron.get("tier"),
        "patronAdtAtTime": patron.get("adt"),
        # interactionEmbedding intentionally omitted — filled by backfill script
    }


# ---------------------------------------------------------------------------
# Patron pool helpers
# ---------------------------------------------------------------------------


def _build_tier_map(patrons: list[dict[str, Any]]) -> dict[str, list[dict[str, Any]]]:
    tier_map: dict[str, list[dict[str, Any]]] = {}
    for p in patrons:
        tier_map.setdefault(p.get("tier", ""), []).append(p)
    return tier_map


def _pick_patron(
    preferred_tiers: list[str],
    tier_map: dict[str, list[dict[str, Any]]],
    all_patrons: list[dict[str, Any]],
) -> dict[str, Any]:
    """70% chance to pick from preferred tier, else uniform over all patrons."""
    if random.random() < 0.7 and preferred_tiers:
        tier = _rand_from(preferred_tiers)
        pool = tier_map.get(tier)
        if pool:
            return _rand_from(pool)
    return _rand_from(all_patrons)


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------


async def main() -> None:
    is_dry_run = "--dry-run" in sys.argv
    db = get_db()
    now = datetime.now(timezone.utc)

    # 1. Load PR agents
    pr_docs = await (
        db[cols.pr_agents]
        .find(
            {},
            projection={
                "prAgentId": 1,
                "name": 1,
                "active": 1,
                "preferredTiers": 1,
            },
        )
        .sort("prAgentId", 1)
        .to_list(length=None)
    )

    if not pr_docs:
        print("No PR agents found. Run `uv run python -m app.seed.seed` first.", file=sys.stderr)
        sys.exit(1)

    # 2. Load up to 300 patrons
    patron_docs = await (
        db[cols.patrons]
        .find({}, projection={"patronId": 1, "tier": 1, "adt": 1})
        .to_list(length=300)
    )

    if not patron_docs:
        print("No patrons found. Run `uv run python -m app.seed.seed` first.", file=sys.stderr)
        sys.exit(1)

    tier_map = _build_tier_map(patron_docs)

    # 3. Generate records per PR
    all_records: list[dict[str, Any]] = []
    for idx, pr in enumerate(pr_docs):
        count = _rand_int(MIN_PER_PR, MAX_PER_PR)
        weights = _type_weights(idx, len(pr_docs))
        for _ in range(count):
            interaction_type = _pick_weighted_type(weights)
            patron = _pick_patron(pr.get("preferredTiers") or [], tier_map, patron_docs)
            all_records.append(_build_record(pr["prAgentId"], patron, interaction_type, now))

    # 4. Preview or insert
    if is_dry_run:
        by_pr: dict[str, dict[str, int]] = {}
        for r in all_records:
            entry = by_pr.setdefault(r["recordedBy"], {})
            entry[r["type"]] = entry.get(r["type"], 0) + 1
        print(
            json.dumps(
                {"dryRun": True, "totalRecords": len(all_records), "byPr": by_pr},
                indent=2,
                default=str,
            )
        )
        return

    await db[cols.patron_interactions].insert_many(all_records)

    # 5. Summary
    type_breakdown: dict[str, int] = {}
    for r in all_records:
        type_breakdown[r["type"]] = type_breakdown.get(r["type"], 0) + 1
    print(
        f"\n✓ Inserted {len(all_records)} interaction records across "
        f"{len(pr_docs)} PR agents."
    )
    print(f"  Type breakdown: {type_breakdown}")
    print(f"  Date range: past {HISTORY_DAYS} days")
    print(
        "\nNext step: run "
        "`uv run python -m app.scripts.backfill_interaction_embeddings` "
        "to generate embeddings for KPI vector search.\n"
    )


async def _entrypoint() -> None:
    try:
        await main()
    finally:
        await close_client()


if __name__ == "__main__":
    asyncio.run(_entrypoint())
